src/main/java/frc/robot/poseestimator/GBPoseEstimator.java

```diff
@@ -0,0 +1,228 @@
+package frc.robot.poseestimator;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Twist2d;
+import edu.wpi.first.math.interpolation.TimeInterpolatableBuffer;
+import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
+import edu.wpi.first.math.kinematics.SwerveDriveWheelPositions;
+import frc.robot.subsystems.GBSubsystem;
+import frc.robot.vision.limelights.GyroAngleValues;
+import frc.robot.vision.limelights.ILimelightFilterer;
+import frc.robot.poseestimator.observations.OdometryObservation;
+import frc.robot.poseestimator.observations.VisionObservation;
+import frc.utils.time.TimeUtils;
+import org.littletonrobotics.junction.Logger;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+
+public class GBPoseEstimator extends GBSubsystem implements IPoseEstimator {
+
+	private final TimeInterpolatableBuffer<Pose2d> odometryPoseInterpolator;
+	private final TimeInterpolatableBuffer<Pose2d> estimatedPoseInterpolator;
+	private final ILimelightFilterer limelightFilterer;
+	private final SwerveDriveKinematics kinematics;
+	private final double[] odometryStandardDeviations;
+	private Pose2d odometryPose;
+	private Pose2d estimatedPose;
+	private SwerveDriveWheelPositions latestWheelPositions;
+	private Rotation2d latestGyroAngle;
+	private Rotation2d headingOffset;
+
+	public GBPoseEstimator(
+		String logPath,
+		ILimelightFilterer limelightFilterer,
+		SwerveDriveKinematics kinematics,
+		SwerveDriveWheelPositions initialWheelPositions,
+		Rotation2d initialGyroAngle,
+		double[] odometryStandardDeviations
+	) {
+		super(logPath);
+		//@formatter:off
+		getVisionPose().ifPresentOrElse(calculatedPose -> {
+			this.odometryPose = calculatedPose;
+			this.estimatedPose = calculatedPose;
+			}, () -> {
+			this.odometryPose = new Pose2d();
+			this.estimatedPose = new Pose2d();
+		});
```

Comment:
3 ways these lines creates a null pointer exception: 
1. `getVisionPose()` is dependent on `this.limelightFilterer`, which has not been initialized yet. 
2. `LimelightFilterer()` is dependent on `LimelightFilterer.getPoseBytimeBlaBla()` which is not (and can't be, since it requires the GBPoseEstimator to be created) initialized in the constructor, and need initialization by `limelightFiltere.setBlaBlaFunction`. 
3. `setBlaBlaFunction` could be called only outside to the constructor (it requires a PoseEstiamtor, means the constructor should be completed first) which breaks everything.
And it makes sense that it breaks everything. The `LimelightFilterer` class is a part of *each* poseestimator -- it uses its inner observations. Each poseestimator would have its own filterer. The only object that should be injected is the MultiLimelight, which can serve several different PoseEstiamtors (or other subsystems too). 

(note that I noticed those exceptions through testing, and fixes for the first two problems could be find in the previous, now reverted, commit in pose-swerve)